src/utils/permissions.ts

```diff
@@ -0,0 +1,321 @@
+  [`role_${ColonyRole.Funding}`]: null,
+  [`role_${ColonyRole.Administration}`]: null,
+};
+
+export const getColonyRolesDatabaseId = (
+  colonyAddress: string,
+  nativeDomainId: number,
+  userAddress: string,
+): string => `${colonyAddress}_${nativeDomainId}_${userAddress}_roles`;
+
+export const getColonyHistoricRolesDatabaseId = (
+  colonyAddress: string,
+  nativeDomainId: number,
+  userAddress: string,
+  blockNumber: number,
+): string => `${colonyAddress}_${nativeDomainId}_${userAddress}_${blockNumber}_roles`;
+
+export const getAllRoleEventsFromTransaction = async (
+  transactionHash: string,
+  colonyAddress: string,
+): Promise<ContractEvent[]> => {
+  const colonyRoleSetEventName = ContractEventsSignatures.ColonyRoleSet.slice(0, ContractEventsSignatures.ColonyRoleSet.indexOf('('));
+  const colonyRecoveryRoleSetName = ContractEventsSignatures.RecoveryRoleSet.slice(0, ContractEventsSignatures.RecoveryRoleSet.indexOf('('));
+
+  const colonyClient = await getCachedColonyClient(colonyAddress);
+  const transactionReceipt = await colonyClient.provider.getTransactionReceipt(transactionHash);
+
+  const events = await Promise.all(
+    transactionReceipt.logs.map(log => mapLogToContractEvent(log, colonyClient.interface)),
+  );
+
+  const filteredEvents = events.filter(event => {
+    if (
+      !event ||
+      !(event.name === colonyRoleSetEventName || event.name === colonyRecoveryRoleSetName)
+    ) {
+      return false;
+    }
+    return true;
+  });
+
+  /*
+   * Typecasting since apparently TS doesn't realize we are actually filtering
+   * to ensure that the Array only contains proper events
+   */
+  return filteredEvents as ContractEvent[];
+};
+
+export const getRolesMapFromEvents = (roleEvents: ContractEvent[], setToNull: boolean = true): Record<string, boolean | null> => {
+  let roleMap = {};
+
+  roleEvents.map(({
+    signature,
+    args: { role, setTo },
+  }) => {
+    /*
+     * @NOTE This logic is begining to get kinda complicated so I'm going to attempt to explain it
+     *
+     * It's role is to format the object key/value pairs for roles (eg: `{ role_1: true }`).
+     *
+     * Since we have a separate event for `RecoveryRoleSet` which has a different format than the rest, check for it, and convert the
+     * role to `0` if we find it.
+     *
+     * Second part (after the || operator check) deals with how we display the "off" side of the role. To keep database consistency, for operational
+     * roles (ie: the ones we use to check against), we check the `false` value (which signifies the permission was taken away) to `null` as that is
+     * the default state of the GraphQL entry. For Actions however, we need to set it to `false` as to distiguish itself, and for us to know what
+     * to display in the Action's text. For this we use `setToNull`, which, when set (on by default), changes the "off" state to null, and when it's
+     * not being set, changes the "off" state to false
```

Comment:
Hard to follow. I think this has something to do with distinguishing between whether a role was set to "off" versus whether it has not been set yet at all?

Reply:
For colony roles / historic roles we only care about if the role is "on" (set to `true`), while everything else that's "off" is set to the `null` which is the default for graphQL (i didn't want to mix both `false` and `null` as to mean "off")

For action roles however, I explicitly need to know if a role was "turned on" (set to `true`) or if it was "turned off" (set to `false`), and I only care about the ones that were changed, not the ones that weren't touched (so that we can generate the action title. ie: role X was added, role Y was removed)